### PR TITLE
[WIP] CES applied to EDMF for parametric uncertainty

### PR DESCRIPTION
**Pull request by odunbar**
## Purpose and Content
To evaluate the efficacy of the E-S stages to an atmospheric calibration problem, where we have data from C-stage

## In this PR:
- zipped data files
- script `uq_for_EDMF.jl`for running CES and creating output files
- script `plot_posterior.jl` for basic plotting of posterior marginals

## TODO: 
- Document example

## Current example outputs: 
1. `exp_name = ent-det-calibration` (takes ~40 mins)

2. `exp_name = ent-det-tked-tkee-stab-calibration` (takes ~80mins)

### CODE REVIEW

File: examples/EDMF_data/uq_for_edmf.jl [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+        outputs = outputs[:, not_all_bad_particles]
+        inputs = inputs[:, not_all_bad_particles]
+        inputs_constrained = inputs_constrained[:, not_all_bad_particles]
+        println("Pruning all-bad particles...")
+        println("removed ", length(not_all_bad_particles)[1] - sum(not_all_bad_particles), " particles.")
+
+        #remove all-bad data dimensions
+        idx_in_bd = (outputs .- min_bd) .> 0 .&& (max_bd .- outputs) .> 0
+        not_all_bad_data_points = vec(any(idx_in_bd, dims = 2))
+        outputs = outputs[not_all_bad_data_points, :]
```
Comment (ilopezgp) [inline examples/EDMF_data/uq_for_edmf.jl:126]: why do we need this step, if you are already filtering outlying particles in L114-L119?

should I understand this as a cropping of dimensions that have structural or model errors outside of the range of the uncertainty given by the truth covariance?

File: examples/EDMF_data/uq_for_edmf.jl [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+    if !isfile(prior_filepath)
+        LoadError("prior file \"prior.jld2\" not found in directory \"" * exp_dir * "/\"")
+    else
+        prior_dict_raw = load(prior_filepath) #using JLD2
+        marginal_priors =
+            ParameterDistributions.ParameterDistribution.(
+                prior_dict_raw["distributions"],
+                prior_dict_raw["constraints"],
+                prior_dict_raw["u_names"],
+            )
```
Comment (ilopezgp) [inline examples/EDMF_data/uq_for_edmf.jl:156]: we will have to change this when we use new data (not now). I am now storing the ParameterDistribution object directly.

File: examples/EDMF_data/uq_for_edmf.jl [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+        )
+    end
+    data_filepath = joinpath(exp_dir, "Diagnostics.nc")
+    if !isfile(data_filepath)
+        LoadError("experiment data file \"Diagnostics.nc\" not found in directory \"" * exp_dir * "/\"")
+    else
+        y_truth = Array(NCDataset(data_filepath).group["reference"]["y_full"]) #ndata
+        truth_cov = Array(NCDataset(data_filepath).group["reference"]["Gamma_full"]) #ndata x ndata
+        # Option (i) get data from NCDataset else get from jld2 files.
+        output_mat = Array(NCDataset(data_filepath).group["particle_diags"]["g_full"]) #nens x ndata x nit
```
Comment (ilopezgp) [inline examples/EDMF_data/uq_for_edmf.jl:82]: Remove last iteration (it has NCDataset fill values)

Comment (ilopezgp) [reply examples/EDMF_data/uq_for_edmf.jl:82]: You can also check whether a value is bad by using:

```
isnan_inf_or_filled(x) = isnan(x) || isinf(x) || x ≈ NC.fillvalue(typeof(x))
```

we use it in TC here https://github.com/CliMA/TurbulenceConvection.jl/blob/main/integration_tests/driver.jl#L76

File: examples/EDMF_data/uq_for_edmf.jl [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+        input_mat = Array(NCDataset(data_filepath).group["particle_diags"]["u"]) #nens x nparam x nit
+        input_constrained_mat = Array(NCDataset(data_filepath).group["particle_diags"]["phi"]) #nens x nparam x nit
+
+        #TODO: Remove mask after testing
```
[the remark this answers is not in the document]
Comment (odunbar) [reply examples/EDMF_data/uq_for_edmf.jl:86]: done

File: examples/EDMF_data/uq_for_edmf.jl [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+        mask = 1:size(y_truth, 1)
+        y_truth = y_truth[mask]
+        truth_cov = truth_cov[mask, mask]
+        output_mat = output_mat[:, mask, :]
+
+        # to be consistent, we wish to go from nens x nparam x nit arrays to nparam x (nit x nens)
```
[the remark this answers is not in the document]
Comment (odunbar) [reply examples/EDMF_data/uq_for_edmf.jl:92]: done

File: examples/EDMF_data/uq_for_edmf.jl [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+
+        outputs = reshape(
+            permutedims(output_mat, (2, 3, 1)),
+            (size(output_mat, 2), size(output_mat, 3) * size(output_mat, 1)),
+        )
+
+        # naive preprocessing of data
+        #remove extremes using truth cov
+        iqr_rows = mapslices(iqr, outputs, dims = 2)
+        med = mapslices(median, outputs, dims = 2)
```
[the remark this answers is not in the document]
Comment (odunbar) [reply examples/EDMF_data/uq_for_edmf.jl:108]: done

File: examples/EDMF_data/uq_for_edmf.jl (hunk 3 above)
Comment (odunbar) [reply examples/EDMF_data/uq_for_edmf.jl:82]: done

**Review by ilopezgp — APPROVED** [review]

**Comment by odunbar** [comment]
I'd like to make a bit of documentation for this example before merge. Could one of you add a little background to start it off? Or could this be obtained by linking to CEDMF or something?

**Review by odunbar — COMMENTED** [review]
LGTM - but I can't approve my own PR.

File: examples/EDMF_data/uq_for_edmf.jl [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+            "/\" not found. Please unzip the compressed file \"" *
+            exp_name *
+            ".zip\" and retry.",
+        )
+    end
+    data_filepath = joinpath(exp_dir, "Diagnostics.nc")
+    if !isfile(data_filepath)
+        LoadError("experiment data file \"Diagnostics.nc\" not found in directory \"" * exp_dir * "/\"")
+    else
+        y_truth = Array(NCDataset(data_filepath).group["reference"]["y_full"]) #ndata
```
Comment (charleskawczynski) [inline examples/EDMF_data/uq_for_edmf.jl:79]: Does this syntax automatically close the opened files?

**Pull request merged**